src/ray/gcs/gcs_server/gcs_health_check_manager.cc

```diff
@@ -124,37 +124,47 @@ void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
     return;
   }
 
-  // Reset the context/request/response for the next request.
-  context_.~ClientContext();
-  new (&context_) grpc::ClientContext();
-  response_.Clear();
+  // grpc context and health check response are dedicated to one single async request.
+  auto context = std::make_shared<grpc::ClientContext>();
+  auto response = std::make_shared<::grpc::health::v1::HealthCheckResponse>();
+
+  // Get the context and response pointer before async call, since the order of function
```

Comment:
I don't understand this comment. But I get what you are trying to do. We can say:

```
Get the raw pointer addresses before the shared_ptrs are moved into the lambda.
```

Reply:
Example: suppose you have a function invocation `f(a, b, c)`, it's undefined behavior which one of a, b, c is passed in first, it's vendor specific implementation.

In this particular case, if lambda is resolved first, shared pointer gets invalidation before it's passed as first argument.

Reply:
Let me add a comment to clarify, I think it's a tricky problem and we need to comment out **why** we do so.

Reply:
I mean, we all know the arg evaluations are non deterministic. But it's not the biggest point here (why we care about args order here?). We need to answer that with: 

> // Get the raw pointer addresses for the Check() call, before the shared_ptrs are moved into the lambda.

Reply:
ok sure, I copied your words